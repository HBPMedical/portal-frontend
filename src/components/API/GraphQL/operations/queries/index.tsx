import createGetExperiment from './Experiments/getExperiment';

export const experimentQueries = {
  getExperiment: createGetExperiment()
};
