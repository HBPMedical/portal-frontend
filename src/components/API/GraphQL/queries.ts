import { gql } from '@apollo/client';
import { coreInfoResult } from './fragments';

export const QUERY_EXPERIMENT = gql`
  ${coreInfoResult}

  query getExperiment($uuid: String!) {
    experiment(uuid: $uuid) {
      name
      uuid
      domain
      datasets
      author
      createdAt
      finishedAt
      viewed
      variables
      coVariables
      shared
      status
      algorithm {
        id
        description
        label
        type
        parameters {
          id
          label
          value
        }
      }
      results {
        ... on GroupsResult {
          groups {
            name
            description
            results {
              ...coreInfoResult
            }
          }
        }
        ...coreInfoResult
      }
    }
  }
`;

export const QUERY_DOMAINS = gql`
  fragment coreGroupInfo on Group {
    id
    label
    description
    groups
    variables
  }

  query listDomains {
    domains {
      id
      label
      description
      datasets {
        id
        label
      }
      variables {
        id
        label
        type
        description
        enumerations {
          id
          label
        }
      }
      rootGroup {
        ...coreGroupInfo
      }
      groups {
        ...coreGroupInfo
      }
    }
  }
`;

export const ADD_TRANSIENT = gql`
  ${coreInfoResult}

  mutation createTransient($data: ExperimentCreateInput!) {
    createExperiment(data: $data, isTransient: true) {
      name
      results {
        ... on GroupsResult {
          groups {
            name
            description
            results {
              ...coreInfoResult
            }
          }
        }
        ...coreInfoResult
      }
    }
  }
`;

export const MUTATE_EXPERIMEMT = gql`
  mutation editExperiment($uuid: String!, $data: ExperimentEditInput!) {
    editExperiment(uuid: $uuid, data: $data) {
      uuid
      name
    }
  }
`;
