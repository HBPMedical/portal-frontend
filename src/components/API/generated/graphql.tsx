import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
const defaultOptions = {};
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  /** A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format. */
  DateTime: any;
};

export type Category = {
  __typename?: 'Category';
  id: Scalars['String'];
  label?: Maybe<Scalars['String']>;
};

export type Domain = {
  __typename?: 'Domain';
  datasets: Array<Category>;
  description?: Maybe<Scalars['String']>;
  groups: Array<Group>;
  id: Scalars['String'];
  label?: Maybe<Scalars['String']>;
  rootGroup: Group;
  variables: Array<Variable>;
};

export type DummyResult = {
  __typename?: 'DummyResult';
  data: Array<Array<Scalars['String']>>;
  groupBy?: Maybe<Scalars['String']>;
  listMax: Array<Scalars['String']>;
  name: Scalars['String'];
};

export type Experiment = {
  __typename?: 'Experiment';
  created_at?: Maybe<Scalars['DateTime']>;
  finished_at?: Maybe<Scalars['DateTime']>;
  results: Array<ResultUnion>;
  title: Scalars['String'];
  update_at?: Maybe<Scalars['DateTime']>;
  uuid?: Maybe<Scalars['String']>;
};

export type ExperimentCreateInput = {
  algorithm: Scalars['String'];
  datasets: Array<Scalars['String']>;
  domain: Scalars['String'];
  filter?: Maybe<Scalars['String']>;
  name: Scalars['String'];
  variables: Array<Scalars['String']>;
};

export type Group = {
  __typename?: 'Group';
  description?: Maybe<Scalars['String']>;
  groups: Array<Group>;
  id: Scalars['String'];
  label?: Maybe<Scalars['String']>;
  variables: Array<Variable>;
};

export type MetaData = {
  __typename?: 'MetaData';
  name: Scalars['String'];
  type: Scalars['String'];
};

export type Mutation = {
  __typename?: 'Mutation';
  createTransient: Experiment;
};

export type MutationCreateTransientArgs = {
  data: ExperimentCreateInput;
};

export type Query = {
  __typename?: 'Query';
  domains: Array<Domain>;
};

export type QueryDomainsArgs = {
  ids?: Maybe<Array<Scalars['String']>>;
};

export type ResultUnion = DummyResult | TableResult;

export type TableResult = {
  __typename?: 'TableResult';
  data: Array<Array<Scalars['String']>>;
  groupBy?: Maybe<Scalars['String']>;
  metadatas: Array<MetaData>;
  name: Scalars['String'];
};

export type Variable = {
  __typename?: 'Variable';
  description?: Maybe<Scalars['String']>;
  enumerations: Array<Category>;
  groups: Array<Group>;
  id: Scalars['String'];
  label?: Maybe<Scalars['String']>;
  type: Scalars['String'];
};

export type ListDomainsQueryVariables = Exact<{ [key: string]: never }>;

export type ListDomainsQuery = {
  __typename?: 'Query';
  domains: Array<{
    __typename?: 'Domain';
    id: string;
    label?: Maybe<string>;
    description?: Maybe<string>;
    datasets: Array<{
      __typename?: 'Category';
      id: string;
      label?: Maybe<string>;
    }>;
    variables: Array<{
      __typename?: 'Variable';
      id: string;
      label?: Maybe<string>;
      type: string;
      description?: Maybe<string>;
      enumerations: Array<{
        __typename?: 'Category';
        id: string;
        label?: Maybe<string>;
      }>;
    }>;
    rootGroup: {
      __typename?: 'Group';
      id: string;
      label?: Maybe<string>;
      description?: Maybe<string>;
      groups: Array<{ __typename?: 'Group'; id: string }>;
      variables: Array<{ __typename?: 'Variable'; id: string }>;
    };
    groups: Array<{
      __typename?: 'Group';
      id: string;
      label?: Maybe<string>;
      description?: Maybe<string>;
      groups: Array<{ __typename?: 'Group'; id: string }>;
      variables: Array<{ __typename?: 'Variable'; id: string }>;
    }>;
  }>;
};

export type CreateTransientMutationVariables = Exact<{
  data: ExperimentCreateInput;
}>;

export type CreateTransientMutation = {
  __typename?: 'Mutation';
  createTransient: {
    __typename?: 'Experiment';
    title: string;
    results: Array<
      | { __typename?: 'DummyResult' }
      | {
          __typename?: 'TableResult';
          groupBy?: Maybe<string>;
          name: string;
          data: Array<Array<string>>;
          metadatas: Array<{
            __typename?: 'MetaData';
            name: string;
            type: string;
          }>;
        }
    >;
  };
};

export const ListDomainsDocument = gql`
  query listDomains {
    domains {
      id
      label
      description
      datasets {
        id
        label
      }
      variables {
        id
        label
        type
        description
        enumerations {
          id
          label
        }
      }
      rootGroup {
        id
        label
        description
        groups {
          id
        }
        variables {
          id
        }
      }
      groups {
        id
        label
        description
        groups {
          id
        }
        variables {
          id
        }
      }
    }
  }
`;

/**
 * __useListDomainsQuery__
 *
 * To run a query within a React component, call `useListDomainsQuery` and pass it any options that fit your needs.
 * When your component renders, `useListDomainsQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useListDomainsQuery({
 *   variables: {
 *   },
 * });
 */
export function useListDomainsQuery(
  baseOptions?: Apollo.QueryHookOptions<
    ListDomainsQuery,
    ListDomainsQueryVariables
  >
) {
  const options = { ...defaultOptions, ...baseOptions };
  return Apollo.useQuery<ListDomainsQuery, ListDomainsQueryVariables>(
    ListDomainsDocument,
    options
  );
}
export function useListDomainsLazyQuery(
  baseOptions?: Apollo.LazyQueryHookOptions<
    ListDomainsQuery,
    ListDomainsQueryVariables
  >
) {
  const options = { ...defaultOptions, ...baseOptions };
  return Apollo.useLazyQuery<ListDomainsQuery, ListDomainsQueryVariables>(
    ListDomainsDocument,
    options
  );
}
export type ListDomainsQueryHookResult = ReturnType<typeof useListDomainsQuery>;
export type ListDomainsLazyQueryHookResult = ReturnType<
  typeof useListDomainsLazyQuery
>;
export type ListDomainsQueryResult = Apollo.QueryResult<
  ListDomainsQuery,
  ListDomainsQueryVariables
>;
export const CreateTransientDocument = gql`
  mutation CreateTransient($data: ExperimentCreateInput!) {
    createTransient(data: $data) {
      title
      results {
        ... on TableResult {
          groupBy
          name
          data
          metadatas {
            name
            type
          }
        }
      }
    }
  }
`;
export type CreateTransientMutationFn = Apollo.MutationFunction<
  CreateTransientMutation,
  CreateTransientMutationVariables
>;

/**
 * __useCreateTransientMutation__
 *
 * To run a mutation, you first call `useCreateTransientMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useCreateTransientMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [createTransientMutation, { data, loading, error }] = useCreateTransientMutation({
 *   variables: {
 *      data: // value for 'data'
 *   },
 * });
 */
export function useCreateTransientMutation(
  baseOptions?: Apollo.MutationHookOptions<
    CreateTransientMutation,
    CreateTransientMutationVariables
  >
) {
  const options = { ...defaultOptions, ...baseOptions };
  return Apollo.useMutation<
    CreateTransientMutation,
    CreateTransientMutationVariables
  >(CreateTransientDocument, options);
}
export type CreateTransientMutationHookResult = ReturnType<
  typeof useCreateTransientMutation
>;
export type CreateTransientMutationResult = Apollo.MutationResult<
  CreateTransientMutation
>;
export type CreateTransientMutationOptions = Apollo.BaseMutationOptions<
  CreateTransientMutation,
  CreateTransientMutationVariables
>;
