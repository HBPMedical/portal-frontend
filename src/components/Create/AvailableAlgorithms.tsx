import * as React from 'react';
import { Button, OverlayTrigger, Popover } from 'react-bootstrap';

import { Algorithm, VariableEntity, AlgorithmParameter } from '../API/Core';
import { ModelResponse } from '../API/Model';
import styled from 'styled-components';
import { Engine } from '../API/Experiment';
interface AvailableAlgorithm extends Algorithm {
  enabled: boolean;
}

const AvailableAlgorithms = ({
  algorithms,
  lookup,
  layout = 'default',
  handleSelectMethod,
  model
}: {
  algorithms: Algorithm[] | undefined;
  layout?: string;
  lookup: (code: string) => VariableEntity;
  handleSelectMethod: (method: Algorithm) => void;
  model: ModelResponse | undefined;
}): JSX.Element => {
  const query = model && model.query;
  const modelVariable =
    (query && query.variables && query.variables.map(v => lookup(v.code))) ||
    [];
  const modelCovariables = [
    ...((query &&
      query.coVariables &&
      query.coVariables.map(v => lookup(v.code))) ||
      []),
    ...((query &&
      query.groupings &&
      query.groupings.map(v => lookup(v.code))) ||
      [])
  ];
  const algorithmEnabled = (
    parameters: AlgorithmParameter[],
    { x, y }: { x: VariableEntity[]; y: VariableEntity[] }
  ): boolean => {
    const checkSelectedVariables = (
      axis: string,
      variables: VariableEntity[]
    ): boolean => {
      const definition = parameters.find(p => p.name === axis);
      if (definition) {
        const isCategorical =
          definition.columnValuesIsCategorical === ''
            ? undefined
            : definition.columnValuesIsCategorical === 'true'
            ? true
            : false;
        // const type = xDefinition.columnValuesSQLType;
        const multiple = definition.valueMultiple;
        const notBlank = definition.valueNotBlank;

        if (isCategorical && !variables.every(c => c.isCategorical)) {
          return false;
        }

        if (isCategorical === false && variables.some(c => c.isCategorical)) {
          return false;
        }

        if (notBlank && variables.length === 0) {
          return false;
        }

        // FIXME: not sure if it MUST or SHOULD be multiple
        // Guessing SHOULD now
        if (!multiple && variables.length > 1) {
          return false;
        }

        return true;
      }

      return true;
    };
    // Independant variable check
    return checkSelectedVariables('x', x) && checkSelectedVariables('y', y);
  };

  const availableAlgorithms: AvailableAlgorithm[] =
    (algorithms &&
      algorithms.map(algorithm => ({
        ...algorithm,
        enabled:
          algorithm.engine === Engine.Workflow
            ? true
            : algorithmEnabled(algorithm.parameters as AlgorithmParameter[], {
                x: modelCovariables,
                y: modelVariable
              })
      }))) ||
    [];

  const variablesHelpMessage = (algorithm: Algorithm): JSX.Element => {
    const message: JSX.Element[] = [];

    const helpFor = (axis: string, term: string): void => {
      const variable = (algorithm.parameters as AlgorithmParameter[]).find(
        p => p.name === axis
      );
      if (variable) {
        if (variable.desc) {
          message.push(
            <p key={`${algorithm.name}-${axis}-desc`}>
              <strong>{term}</strong>: {variable.desc}
            </p>
          );
        } else {
          message.push(
            <p key={`${algorithm.name}-${axis}-desc`}>
              <strong>{term}</strong>
            </p>
          );
        }

        const isCategorical =
          variable.columnValuesIsCategorical === ''
            ? undefined
            : variable.columnValuesIsCategorical === 'true'
            ? true
            : false;

        let multipleconstraint = '';
        if (!variable.valueMultiple) {
          multipleconstraint = ', one var max';
        } else {
          multipleconstraint = ', multiple vars';
        }

        if (isCategorical) {
          message.push(
            <p key={`${algorithm.name}-${axis}-1`}>
              - hint: should be multinominal{multipleconstraint}
            </p>
          );
        } else if (isCategorical === false) {
          message.push(
            <p key={`${algorithm.name}-${axis}-2`}>
              - hint: should be continous{multipleconstraint}
            </p>
          );
        } else if (!isCategorical) {
          message.push(
            <p key={`${algorithm.name}-${axis}-3`}>
              - hint: can be either multinominal or continuous
              {multipleconstraint}
            </p>
          );
        }
      }
    };

    helpFor('y', 'Variable (dependant)');
    helpFor('x', 'Covariable (independant)');

    return <>{message}</>;
  };

  const Container = styled.div`
    var::after {
      content: ', ';
    }

    var::last-child::after {
      content: '';
    }

    p {
      margin: 0;
      padding: 0;
      border: 1px solid transparent;
    }
  `;

  // const PLongMethodName = styled.p`
  //   margin: 0 0 8px 0;
  //   overflow: wrap;
  //   width: 220px;
  //   word-wrap: break-word;
  //   display: inline-block;
  // `;

  return (
    <Container>
      {availableAlgorithms.map(algorithm => (
        <OverlayTrigger
          key={algorithm.name}
          placement="left"
          rootClose={false}
          overlay={
            <Popover id={`tooltip-${algorithm.name}`}>
              <h4>{algorithm.name}</h4>
              <p>{algorithm.desc}</p>
              {algorithm.engine === Engine.Workflow
                ? ''
                : variablesHelpMessage(algorithm)}
            </Popover>
          }
        >
          {layout !== 'inline' ? (
            <div>
              {algorithm.enabled && (
                <Button
                  key={algorithm.name}
                  bsStyle="link"
                  // ts lint:disable-next-line jsx-no-lambda
                  onClick={(): void => handleSelectMethod(algorithm)}
                  disabled={!algorithm.enabled}
                  style={{
                    color: algorithm.enabled ? '#03a9f4' : 'gray',
                    padding: 0,
                    textTransform: 'none'
                  }}
                >
                  {algorithm.name}
                </Button>
              )}
              {!algorithm.enabled && (
                <p
                  key={algorithm.name}
                  style={{
                    color: algorithm.enabled ? '#03a9f4' : 'gray',
                    padding: 0,
                    textTransform: 'none'
                  }}
                >
                  {algorithm.name}
                </p>
              )}
            </div>
          ) : (
            <var
              key={algorithm.name}
              style={{
                color: algorithm.enabled ? '#03a9f4' : 'gray',
                padding: 0,
                textTransform: 'none'
              }}
            >
              {algorithm.name}
            </var>
          )}
        </OverlayTrigger>
      ))}
    </Container>
  );
};
export default AvailableAlgorithms;
