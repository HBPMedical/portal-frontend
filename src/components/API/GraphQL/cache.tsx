import { InMemoryCache, makeVar } from '@apollo/client';
import { SessionState } from '../../../utilities/types';
import { Configuration, Domain, Experiment, User } from './types.generated';

export const initialExperiment: Experiment = {
  id: '',
  algorithm: {
    id: '',
    parameters: []
  },
  datasets: [],
  domain: '',
  name: 'New experiment',
  shared: false,
  viewed: false,
  variables: []
};

export const initialUser = {
  id: '',
  username: 'anonymous'
};

export const initialConfig: Configuration = {
  connectorId: 'default',
  hasGalaxy: false,
  version: ''
};

export const sessionStateVar = makeVar<SessionState>(SessionState.INIT);
export const configurationVar = makeVar<Configuration>(initialConfig);
export const zoomNodeVar = makeVar<string | undefined>(undefined);
export const domainsVar = makeVar<Domain[]>([]);
export const currentUserVar = makeVar<User | undefined>(undefined);
export const selectedVariableVar = makeVar<string | undefined>(undefined);
export const selectedDomainVar = makeVar<Domain | undefined>(undefined);
export const selectedExperimentVar = makeVar<Experiment | undefined>(undefined);
export const draftExperimentVar = makeVar<Experiment>(initialExperiment);
export const userVar = makeVar<typeof initialUser>(initialUser);

export const cache: InMemoryCache = new InMemoryCache({
  possibleTypes: {
    // https://github.com/apollographql/apollo-client/issues/7050
    ResultUnion: [
      'RawResult',
      'GroupsResult',
      'TableResult',
      'HeatMapResult',
      'LineChartResult'
    ]
  },
  typePolicies: {
    // see https://www.apollographql.com/docs/react/local-state/managing-state-with-field-policies/
    // should be calculated in the gateway this is temporary as there is no session in the gateway at the moment
    Experiment: {
      fields: {
        isOwner: {
          read(_, { readField }): boolean {
            const author = readField<Experiment['author']>('createdBy');
            return userVar().username === author?.username;
          }
        }
      }
    }
  }
});
