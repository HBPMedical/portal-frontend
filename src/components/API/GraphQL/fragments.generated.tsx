import * as Types from './types.generated';

import { gql } from '@apollo/client';
export type CoreInfoResult_GroupsResult_Fragment = { __typename?: 'GroupsResult' };

export type CoreInfoResult_HeatMapResult_Fragment = { __typename?: 'HeatMapResult', name: string, matrix: Array<Array<number>>, xAxis?: Types.Maybe<{ __typename?: 'ChartAxis', label?: Types.Maybe<string>, categories?: Types.Maybe<Array<string>> }>, yAxis?: Types.Maybe<{ __typename?: 'ChartAxis', label?: Types.Maybe<string>, categories?: Types.Maybe<Array<string>> }> };

export type CoreInfoResult_LineChartResult_Fragment = { __typename?: 'LineChartResult', name: string, xAxis?: Types.Maybe<{ __typename?: 'ChartAxis', label?: Types.Maybe<string> }>, yAxis?: Types.Maybe<{ __typename?: 'ChartAxis', label?: Types.Maybe<string> }>, lines: Array<{ __typename?: 'LineResult', label: string, x: Array<number>, y: Array<number>, type?: Types.Maybe<Types.LineType>, extraLineInfos?: Types.Maybe<Array<{ __typename?: 'ExtraLineInfo', label: string, values: Array<string> }>> }> };

export type CoreInfoResult_RawResult_Fragment = { __typename?: 'RawResult', rawdata: any };

export type CoreInfoResult_TableResult_Fragment = { __typename?: 'TableResult', name: string, data: Array<Array<string>>, headers: Array<{ __typename?: 'Header', name: string, type: string }> };

export type CoreInfoResultFragment = CoreInfoResult_GroupsResult_Fragment | CoreInfoResult_HeatMapResult_Fragment | CoreInfoResult_LineChartResult_Fragment | CoreInfoResult_RawResult_Fragment | CoreInfoResult_TableResult_Fragment;

export const CoreInfoResultFragmentDoc = gql`
    fragment coreInfoResult on ResultUnion {
  ... on RawResult {
    rawdata
  }
  ... on TableResult {
    name
    data
    headers {
      name
      type
    }
  }
  ... on HeatMapResult {
    name
    matrix
    xAxis {
      label
      categories
    }
    yAxis {
      label
      categories
    }
  }
  ... on LineChartResult {
    name
    xAxis {
      label
    }
    yAxis {
      label
    }
    lines {
      label
      x
      y
      extraLineInfos {
        label
        values
      }
      type
    }
  }
}
    `;