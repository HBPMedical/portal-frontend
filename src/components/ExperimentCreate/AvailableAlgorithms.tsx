import React from 'react';
import { Card, OverlayTrigger, Popover } from 'react-bootstrap';
import styled from 'styled-components';
import { useListAlgorithmsQuery } from '../API/GraphQL/queries.generated';
import {
  Algorithm,
  Experiment,
  Variable,
  VariableParameter,
} from '../API/GraphQL/types.generated';
import Loader from '../UI/Loader';

const Container = styled.div`
  line-height: 15px;

  .algorithms {
    .algorithm {
      cursor: default;
      color: grey;
      &.selected {
        font-weight: bold;
      }
    }
    &.clickable {
      .algorithm.enabled {
        cursor: pointer;
      }
    }
    &.vertical {
      display: flex;
      flex-direction: column;
      .algorithm {
        margin-bottom: 10px;
        font-size: 0.9rem;
        &.enabled {
          color: #2b33e9;
          font-weight: bold;
          &:hover {
            text-decoration: underline;
          }
        }
      }
    }
    &.horizontal {
      .algorithm {
        font-size: 0.8rem;
        font-style: italic;
        &.enabled {
          color: #2b33e9;
          font-weight: bold;
        }
        &::after {
          content: ', ';
        }
        &:last-child::after {
          content: '';
        }
      }
    }
  }
`;

const PopoverContainer = styled(Popover)`
  .card {
    margin-bottom: 0px;
  }
`;

const checkValidity = (
  validator: VariableParameter | null | undefined,
  vars: Variable[]
): boolean => {
  if (!validator) return true;
  if (vars.length === 0) return !validator.isRequired;

  const filtered = vars.filter(
    (v, i) =>
      (!validator.allowedTypes ||
        validator.allowedTypes?.includes(v.type ?? '')) &&
      (validator.hasMultiple || i === 0)
  );

  return filtered.length === vars.length;
};

type Props = {
  experiment: Experiment;
  selectedAlgorithm?: Algorithm;
  direction?: 'horizontal' | 'vertical';
  handleSelect?: (algo: Algorithm) => void;
  listVariables: Variable[];
};

export const AvailableAlgorithms = ({
  direction = 'horizontal',
  handleSelect,
  experiment,
  listVariables,
  selectedAlgorithm,
}: Props) => {
  const isClickable = !!handleSelect;
  const { data, loading } = useListAlgorithmsQuery();
  const variables = experiment.variables
    .map((id) => listVariables.find((v) => v.id === id))
    .filter((v) => v)
    .map((v) => v as Variable);
  const coVariables = (
    experiment.coVariables?.map((id) =>
      listVariables.find((v) => v.id === id)
    ) ?? []
  )
    .filter((v) => v)
    .map((v) => v as Variable);

  const algorithms =
    data?.algorithms
      .map((algo) => {
        return {
          ...algo,
          label: (algo.label ?? algo.id).trim(),
          isEnabled:
            checkValidity(algo.variable, variables) &&
            checkValidity(algo.coVariable, coVariables),
        };
      })
      .sort((a, b) => a.label.localeCompare(b.label)) ?? [];

  if (loading) return <Loader />;

  return (
    <Container>
      <div
        className={`algorithms ${direction} ${isClickable ? 'clickable' : ''}`}
      >
        {algorithms.map((algo) => (
          <OverlayTrigger
            key={algo.id}
            placement="left"
            rootClose={false}
            overlay={
              <PopoverContainer id={`tooltip-${algo.label}`}>
                <Card>
                  <Card.Body>
                    <h5>{algo.label}</h5>
                    <p>{algo.description}</p>
                    {algo.variable && (
                      <p>
                        <strong>Variable (dependant)</strong>:{' '}
                        {algo.variable.hint}
                        {algo.variable.allowedTypes && (
                          <>
                            <br />
                            <small
                              color="grey"
                              className="font-italic text-muted"
                            >
                              Allowed types :{' '}
                              {algo.variable.allowedTypes.join(', ')}
                            </small>
                          </>
                        )}
                      </p>
                    )}
                    {algo.coVariable && (
                      <p>
                        <strong>Covariate (independant)</strong>:{' '}
                        {algo.coVariable.hint ?? 'N/A'}
                        {algo.coVariable.allowedTypes && (
                          <>
                            <br />
                            <small
                              color="grey"
                              className="font-italic text-muted"
                            >
                              Allowed types :{' '}
                              {algo.coVariable.allowedTypes.join(', ')}
                            </small>
                          </>
                        )}
                      </p>
                    )}
                  </Card.Body>
                </Card>
              </PopoverContainer>
            }
          >
            <span
              className={`algorithm ${
                algo.isEnabled ? 'enabled' : 'disabled'
              } ${
                selectedAlgorithm && selectedAlgorithm.id === algo.id
                  ? 'selected'
                  : ''
              }`}
              key={algo.id}
              onClick={() => {
                if (handleSelect && algo.isEnabled) {
                  handleSelect(algo as Algorithm);
                }
              }}
            >
              {algo.label ?? algo.id}
            </span>
          </OverlayTrigger>
        ))}
      </div>
    </Container>
  );
};

export default AvailableAlgorithms;
