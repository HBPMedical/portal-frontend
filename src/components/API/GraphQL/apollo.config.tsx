import { ApolloClient, ApolloLink, HttpLink } from '@apollo/client';
import { onError } from '@apollo/client/link/error';
import { graphQLURL } from '../RequestURLS';
import { cache } from './cache';
import config from '../RequestHeaders';

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({ message, locations, path }) =>
      console.log(
        `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
      )
    );
  if (networkError) console.log(`[Network error]: ${networkError}`);
});

export const apolloClient = new ApolloClient({
  link: ApolloLink.from([
    errorLink,
    new HttpLink({ uri: graphQLURL, credentials: 'include' })
  ]),
  headers: {
    ...config.options?.headers,
    accept: 'application/json, text/plain, */*'
  },
  cache: cache,
  credentials: 'include'
});
