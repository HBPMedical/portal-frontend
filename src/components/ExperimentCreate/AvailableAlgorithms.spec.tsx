import { MockedProvider } from '@apollo/client/testing';
import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { initialExperiment } from '../API/GraphQL/cache';
import { QUERY_LIST_ALGORITHMS } from '../API/GraphQL/queries';
import { namedOperations } from '../API/GraphQL/queries.generated';
import {
  Algorithm,
  Experiment,
  Variable
} from '../API/GraphQL/types.generated';
import AvailableAlgorithms from './AvailableAlgorithms';

const exp: Experiment = {
  ...initialExperiment,
  variables: ['var1']
};

const algo: Algorithm = {
  id: 'dummy-id',
  type: 'dummy-type',
  label: 'Dummy algo',
  description: 'test',
  __typename: 'Algorithm',
  hasFormula: false,
  variable: {
    __typename: 'VariableParameter',
    hint: 'test',
    isRequired: true,
    allowedTypes: ['real'],
    hasMultiple: false
  },
  coVariable: {
    __typename: 'VariableParameter',
    hint: 'test',
    isRequired: true,
    allowedTypes: [], // accept no types
    hasMultiple: false
  },
  parameters: []
};

const algoOneVar: Algorithm = {
  id: 'dummy-id2',
  type: 'dummy-type',
  label: 'Dummy algo2',
  description: 'test2',
  __typename: 'Algorithm',
  hasFormula: false,
  variable: {
    __typename: 'VariableParameter',
    hint: 'test',
    isRequired: false,
    allowedTypes: ['integer'],
    hasMultiple: false
  },
  coVariable: null,
  parameters: []
};

const listVars: Variable[] = [
  { id: 'var1', label: 'Variable 1', type: 'integer' }
];

const pipe = (algorithms = [algo]) => ({
  request: {
    operationName: namedOperations.Query.listAlgorithms,
    query: QUERY_LIST_ALGORITHMS
  },
  result: {
    data: {
      algorithms
    }
  }
});

const mocks = [pipe([]), pipe([])];

const renderComponent = (
  exp: Experiment,
  vars: Variable[],
  data = mocks,
  selectedAlgorithm?: Algorithm
) => {
  return render(
    <MockedProvider mocks={data} addTypename={true}>
      <AvailableAlgorithms
        experiment={exp}
        listVariables={vars}
        selectedAlgorithm={selectedAlgorithm}
      />
    </MockedProvider>
  );
};

describe('Available Algorithms component', () => {
  describe('when there is no algorithm', () => {
    it('should display nothing', async () => {
      const { asFragment, getByText } = renderComponent(exp, []);

      expect(getByText('loading...')).not.toBeNull();

      await waitFor(async () => {
        await new Promise(resolve => setTimeout(resolve, 100));
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there are algorithms', () => {
    it('should display one algorithm enabled', async () => {
      const { asFragment } = renderComponent(exp, listVars, [
        pipe([algo, algoOneVar])
      ]);

      await waitFor(async () => {
        await new Promise(resolve => setTimeout(resolve, 100));
      });

      expect(asFragment()).toMatchSnapshot();
    });

    it('should display one algorithm selected', async () => {
      const { asFragment } = renderComponent(
        exp,
        listVars,
        [pipe([algo, algoOneVar])],
        algoOneVar
      );

      await waitFor(async () => {
        await new Promise(resolve => setTimeout(resolve, 100));
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
