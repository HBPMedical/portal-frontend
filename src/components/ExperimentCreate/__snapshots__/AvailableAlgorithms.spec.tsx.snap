// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Available Algorithms component when there are algorithms should display one algorithm enabled 1`] = `
<DocumentFragment>
  <div
    class="sc-hKMtZM eFzUpx"
  >
    <div
      class="algorithms horizontal "
    >
      <span
        class="algorithm disabled "
      >
        Dummy algo
      </span>
      <span
        class="algorithm enabled "
      >
        Dummy algo2
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Available Algorithms component when there are algorithms should display one algorithm selected 1`] = `
<DocumentFragment>
  <div
    class="sc-hKMtZM eFzUpx"
  >
    <div
      class="algorithms horizontal "
    >
      <span
        class="algorithm disabled "
      >
        Dummy algo
      </span>
      <span
        class="algorithm enabled selected"
      >
        Dummy algo2
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Available Algorithms component when there is no algorithm should display nothing 1`] = `
<DocumentFragment>
  <div
    class="sc-hKMtZM eFzUpx"
  >
    <div
      class="algorithms horizontal "
    />
  </div>
</DocumentFragment>
`;
